src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSource.java

```diff
@@ -285,8 +290,85 @@ protected void retrieve(@CheckForNull SCMSourceCriteria criteria, SCMHeadObserve
                                " Check the configuration before running this job again.");
                 return;
             }
-            getFullyInitializedGitSCMSource().accessibleRetrieve(criteria, observer, event, listener);
+
+            if (hasLegacyTraits()) {
+                doRetrieveLegacy(criteria, observer, event, listener);
+            }
+
+            doRetrieve(criteria, observer, event, listener);
```

Comment:
Instead of working around and attempting to "pass" data through to the `GitSCMSource#retrieve` method, I found that we can simply run our own retrieve after the previous one has finished. This is only possible if we are to use a PR discovery trait as opposed to a filter trait as the latter needs access to the heads fetch in the retrieve.